mirgecom/float_comparisons.py

```diff
@@ -0,0 +1,70 @@
+""":mod:`mirgecom.float_comparisons` provides comparisons for float-valued arrays.
+
+Comparison Functions
+^^^^^^^^^^^^^^^^^^^^
+.. autofunction:: componentwise_norm
+.. autofunction:: componentwise_err
+.. autofunction:: within_tol
+"""
+
+__copyright__ = """
+Copyright (C) 2021 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+from mirgecom.fluid import ConservedVars
+from pytools.obj_array import obj_array_vectorize
+import numpy as np
+
+
+def componentwise_norm(discr, a):
+    """Calculate a component-wise infinity norm."""
+    if isinstance(a, ConservedVars):
+        return componentwise_norm(discr, a.join())
+    return obj_array_vectorize(lambda b: discr.norm(b, np.inf), a)
```

Comment:
This should use array containers.

Reply:
I think I switched over, but there's some subtle differences between `map_array_container` and `obj_array_vectorize` that I don't entirely understand so I'm not sure if the way I do it by (33e37cc) is the most efficient.

Reply:
So, I think what @inducer is getting at is that `ConservedVars` is an array container, so in theory you should be able to replace the whole `componentwise_norm` implementation with `return map_array_container(lambda b: discr.norm(b, order), a)`. However, if you subsequently call this version on a `ConservedVars`, you'll get a `ConservedVars` back (containing the norm of each component), so I think you'll need to do an additional array container reduction inside `within_tol`. I don't have much hands-on experience with array container reductions, but it looks like there is a `rec_map_reduce_array_container` that might do the job.

Reply:
Oh, that's a good point. Thank you so much. There's some weird differences that I don't completely understand, but they make `map_array_container` not a drop-in replacement for `obj_array_vectorize`, though. I'll do some more testing and see what I can do.